realm/app-manager/src/manager.rs

```diff
@@ -0,0 +1,272 @@
+
+use crate::app::Application;
+use crate::config::{Config, KeySealingType, LauncherType};
+use crate::key::{dummy::DummyKeySealing, KeySealing};
+use crate::launcher::handler::ApplicationHandlerError;
+use crate::launcher::LauncherError;
+use crate::launcher::{dummy::DummyLauncher, Launcher};
+use crate::util::os::{reboot, RebootAction};
+use crate::Error as MainError;
+
+use super::Result;
+pub type ProtocolResult<T> = std::result::Result<T, ProtocolError>;
+
+#[derive(Debug, Error)]
+pub enum ManagerError {
+    #[error("Invalid launcher")]
+    InvalidLauncher(),
+
+    #[error("Failed to join the provisioning thread")]
+    ProvisionJoinError(#[source] JoinError),
+
+    #[error("Vsock connection error")]
+    VsockConnectionError(#[source] std::io::Error),
+
+    #[error("Framed json error")]
+    FramedJsonError(#[from] JsonFramedError),
+
+    #[error("Invalid message received")]
+    InvalidMessage(),
+}
+
+pub struct Manager {
+    config: Config,
+    apps: HashMap<Uuid, Application>,
+    conn: JsonFramed<VsockStream, Request, Response>,
+}
+
+impl Manager {
+    pub async fn new(config: Config) -> Result<Self> {
+        let vsock = VsockStream::connect(VsockAddr::new(VMADDR_CID_HOST, config.vsock_port))
+            .await
+            .map_err(ManagerError::VsockConnectionError)?;
+        info!("Connected to warden daemon");
+
+        Ok(Self {
+            config,
+            apps: HashMap::new(),
+            conn: JsonFramed::new(vsock),
+        })
+    }
+
+    fn make_launcher(&self) -> Result<Box<dyn Launcher + Send + Sync>> {
+        match self.config.launcher {
+            LauncherType::Dummy => Ok(Box::new(DummyLauncher::new())),
+        }
+    }
+
+    fn make_keyseal(&self) -> Result<Box<dyn KeySealing + Send + Sync>> {
+        match self.config.keysealing {
+            KeySealingType::Dummy => Ok(Box::new(DummyKeySealing::new(vec![0x11, 0x22, 0x33]))),
+        }
+    }
+
+    async fn recv_msg(&mut self) -> Result<Request> {
+        let msg = self
+            .conn
+            .recv()
+            .await
+            .map_err(ManagerError::FramedJsonError)?;
+
+        Ok(msg)
+    }
+
+    async fn send_msg(&mut self, resp: Response) -> Result<()> {
+        self.conn
+            .send(resp)
+            .await
+            .map_err(ManagerError::FramedJsonError)?;
+
+        Ok(())
+    }
+
+    async fn recv_provision_info(&mut self) -> Result<Vec<ApplicationInfo>> {
+        let msg = self.recv_msg().await?;
+
+        if let Request::ProvisionInfo(infos) = msg {
+            Ok(infos)
+        } else {
+            error!("Provision info not received, got: {:?}", msg);
+
+            Err(ManagerError::InvalidMessage().into())
+        }
+    }
+
+    async fn report_provision_success(&mut self) -> Result<()> {
+        self.conn
+            .send(Response::Success())
+            .await
+            .map_err(ManagerError::FramedJsonError)?;
+
+        Ok(())
+    }
```

Comment:
There is also `Response::ProvisioningFinished` which in fact isn't used anywhere in the code, did you consider using this request here? You can also delete previously mentioned response and keep this part of the code unchanged (which might be even the best option).

Reply:
Yeah, you are right, I removed it.